### PR TITLE
Output Config

### PR DESCRIPTION
**Pull request by chleh**
This PR:
* adds a class for writing basic PVD files
* adds a class that handles output configuration

Output files now have names like `square_1e4_pcs_0_ts_1_t_1.000000.vtu`.
The output is configured by a snippet like the following in the prj file:
```
    <output>
        <type>VTK</type>
        <prefix>line_1</prefix>
        <timesteps>
            <pair><repeat> 10 </repeat><each_steps>  1 </each_steps></pair>
            <pair><repeat> 10 </repeat><each_steps>  1 </each_steps></pair>
            <pair><repeat> 10 </repeat><each_steps>  1 </each_steps></pair>
            <pair><repeat> 10 </repeat><each_steps>  1 </each_steps></pair>
            <pair><repeat> 10 </repeat><each_steps>  1 </each_steps></pair>
            <pair><repeat> 10 </repeat><each_steps>  1 </each_steps></pair>
            <pair><repeat>  1 </repeat><each_steps> 10 </each_steps></pair>
        </timesteps>
    </output>
```
where the `<timesteps>` part is optional and defaults to doing output in every timestep.

The input files changed accordingly: E.g. `<file>line_1</file>` → `<prefix>line_1</prefix>`.

### CODE REVIEW

**Comment by chleh** [comment]
Looks like g++ 4.8's lib is too old. ^^ Gonna workaround.

File: ProcessLib/Output.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    auto const pcs = 0u; // TODO fix
+
+    std::string const output_file_name =
+            _output_file_prefix + "_pcs_" + std::to_string(pcs)
+            + "_ts_" + std::to_string(timestep)
+            + "_t_"  + std::to_string(t)
+            + ".vtu";
+    DBUG("output to %s", output_file_name.c_str());
+    // auto const& p = _processes[pcs];
```
Comment (chleh) [inline ProcessLib/Output.cpp]: TODO remove.

File: FileIO/VtkIO/PVDFile.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+PVDFile::PVDFile(const std::string &fn)
+    : _fh{fn.c_str()}
+{
+    if (!_fh) {
+        ERR("could not open file `%s'", fn.c_str());
+        std::abort();
+    }
+
+    _fh << std::setprecision(16);
```
Comment (TomFischer) [inline FileIO/VtkIO/PVDFile.cpp]: Why "16"? Maybe you could use `std::numeric_limits<double>::digits10`?

File: FileIO/VtkIO/PVDFile.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <fstream>
+#include <iomanip>
+#include "PVDFile.h"
+
+#include "logog/include/logog.hpp"
```
Comment (chleh) [inline FileIO/VtkIO/PVDFile.cpp]: TODO use <>

**Comment by chleh** [comment]
Fixing PETScVector in this PR needs too many changes. So I removed one commit.

File: Applications/ApplicationsLib/ProjectData.h [hunk 4]
```diff
@@ -63,7 +64,8 @@ class ProjectData final
 	/// The additional  path is used to find files referenced in the
 	/// configuration.
 	ProjectData(BaseLib::ConfigTree const& config_tree,
-	            std::string const& path);
+	            std::string const& path,
+	            std::string const& output_directory);
```
Comment (TomFischer) [inline Applications/ApplicationsLib/ProjectData.h]: - Maybe `path` should be renamed to `input_path` or `input_directory` to reflect better that this variable is used for input in contrast to the `output_directory`? 
- I would prefer `output_path` instead of `output_directory`.
- For consistency I would either document all parameters or none. At the moment `path` is documented, `output_directory` is not.

File: FileIO/VtkIO/PVDFile.cpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2015, OpenGeoSys Community (http://www.opengeosys.org)
```
Comment (TomFischer) [inline FileIO/VtkIO/PVDFile.cpp]: 2012-2016

**Comment by norihiro-w** [comment]
General comments
- why create vtk file for each process? This is inconvenient and using more disk space
- about time information in vtk file names: what is format? Scientific or fixed? How many digits? Can Paraview recognize their sequence without a PVD file?
- It's convenient if a PVD file is updated in a valid form at each output. You don't want to check results during simulations?

**Comment by norihiro-w** [comment]
For config, I would use repeat instead of count

**Comment by chleh** [comment]
> General comments
- why create vtk file for each process? This is inconvenient and using more disk space

When writing only one VTK file there could be name clashes for different outputs, especially with secondary process variables. Having one file per process does not have such problems. Actually I don't even know, if this approach will work for multiple processes. Currently there is only one. Probably with the presented implementation every two processes on the same mesh will effectively write almost identical VTK files (I won't fix that in this PR!). However different processes could live on different meshes. So probably each process should have their own output.

> - about time information in vtk file names: what is format? Scientific or fixed? How many digits? 

`std::to_string` generates fixed output with 6 decimal places. Cf. http://en.cppreference.com/w/cpp/string/basic_string/to_string and http://en.cppreference.com/w/cpp/io/c/fprintf

> Can Paraview recognize their sequence without a PVD file?

I don't know. But why can't you simply open the PVD file? (Probably not during simulation...)

> - It's convenient if a PVD file is updated in a valid form at each output. You don't want to check results during simulations?

I think that can be done in a subsequent PR. This PR is meant as a first attempt to encapsulate output control in a separate class. If you need more, nobody keeps you from extending this functionality.

File: Applications/ApplicationsLib/ProjectData.h (hunk 4 above)
Comment (chleh) [inline Applications/ApplicationsLib/ProjectData.h]: I prefer the suffix `_directory`, because the arguments refer to directories.
From `..._path` it is not directly visible if files or dirs are meant.

**Comment by norihiro-w** [comment]
First of all, I'm not familiar with current implementation (in simulation part) any more so that I can comment only on what features we need and what are missing. Depending on the amount of works needed to implement the requirement, you guys can decide if it's worth to implement them in this PR or separate PRs. The followings are my reply to each comment you gave above:
- You are thinking output is actually dependent on a mesh, which I also agree. then why not having a such interface instead of process-dependent? Since the two approaches need different code structure, it's better if we implement it as soon as possible before another process type is coming in. Concerning variable names, we just need to give different names (e.g. (pcs name)_(var name)) or ask a user to name them. 
- I would remove time information from the file names. Though exceeding 6 decimal places can merely happen, there is no convincing reason to include such an unnecessary limitation which we many need to fix in the future. Having a time step number in the file name is enough to distinguish files from different times. and anyway time information is stored in a PVD file.
- It helps when you don't have a PVD or when you lost it, or when it is broken.
- That can be done in a separate PR if it will not change code structure or interface a lot.

**Comment by chleh** [comment]
I'd like to make as few changes as possible to this PR now, because there are only today and tomorrow left before my vacation. Sure, that's a lame excuse. But (i) I need this PR for the upcoming TES process, and (ii) I won't have time in april for doing such things.

Your comments are important for the long run, but currently I simply don't have time.

> I would remove time information from the file names. [...] Having a time step number in the file name is enough to distinguish files from different times. [...]

I'd keep it there. Especially since the filenames are already unique without the time information.
Adding time information makes it easy to see from the file name with which time you are dealing. I think, it's a rather handy feature when handling output files.

> It helps when you don't have a PVD or when you lost it, or when it is broken.

What? The time in the file name? Then I agree.

**Comment by norihiro-w** [comment]
Sorry I don't totally get it. Maybe thre is something I don't know about current development way. In general, if you need some code to run your application or you don't have time for improvement, why do you need to make a pull request? You just need to stick with your branch, don't you? If we merge PRs because of these reasons, there is no point to review the codes.

**Comment by chleh** [comment]
The merge of the TES process into master has been planned since last september. Furthermore: The inner parts of OGS6 are constantly moving. I don't have the time to maintain my branches all the time.

Furthermore: I think it is much better to include a simple implementation of some feature early and extend it later on, than to wait until that feature is perfect. If we would do the latter, OGS6 maybe wouldn't have moved in any direction in the last six months.

Also: I think the quality of my code is not bad. It is just not feature-complete.

**Comment by norihiro-w** [comment]
> The merge of the TES process into master has been planned since last september. Furthermore: The inner parts of OGS6 are constantly moving. I don't have the time to maintain my branches all the time.

That's your problem and doesn't give any reason to merge PRs because you don't have time or you need it asap. 

> Furthermore: I think it is much better to include a simple implementation of some feature early and extend it later on, than to wait until that feature is perfect. If we would do the latter, OGS6 maybe wouldn't have moved in any direction in the last six months.

This way of answering is much better than above. Simpler is better, but too simple is not god.

> Also: I think the quality of my code is not bad. It is just not feature-complete.

This is what code reviewers decide for each PR.

**Comment by norihiro-w** [comment]
@endJunction  to be discussed

**Comment by norihiro-w** [comment]
> I'd keep it there. Especially since the filenames are already unique without the time information.
Adding time information makes it easy to see from the file name with which time you are dealing. I think, it's a rather handy feature when handling output files.

Go ahead if you insist it. 

> What? The time in the file name? Then I agree.

What? I don't understand your reply.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h [hunk 6]
```diff
@@ -29,7 +29,7 @@ template<typename Matrix, typename Vector>
 class UncoupledProcessesTimeLoop
 {
 public:
-    bool loop(ProjectData& project, std::string const& outdir);
+    bool loop(ProjectData& project);
```
Comment (norihiro-w) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: Doesn't it makes sense to pass project in constructor?

File: FileIO/VtkIO/PVDFile.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+void PVDFile::addVTUFile(const std::string &fn, double timestep)
+{
+    _fh << "    <DataSet timestep=\"" << timestep << "\" group=\"\" part=\"0\" file=\"" << fn << "\"/>\n";
+}
+
+PVDFile::~PVDFile()
+{
+    _fh << "  </Collection>\n</VTKFile>\n";
```
Comment (norihiro-w) [inline FileIO/VtkIO/PVDFile.cpp]: I don't think having an output in destructor is a good idea. Destructor should be used only for finalizing resources. Moreover, it's not an intuitive interface. A user of this class has to delete an object to get a complete output.

File: FileIO/VtkIO/PVDFile.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <fstream>
+#include <iomanip>
+#include <limits>
+#include <logog/include/logog.hpp>
+
+#include "PVDFile.h"
```
Comment (norihiro-w) [inline FileIO/VtkIO/PVDFile.cpp]: should be included at the top 

see http://ufz.github.io/styleguide/cppguide.xml?showone=Names_and_Order_of_Includes#Names_and_Order_of_Includes

File: ProcessLib/Output.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+ */
+
+#include<cassert>
+#include<fstream>
+#include<vector>
+
+#include <logog/include/logog.hpp>
+
+#include "BaseLib/FileTools.h"
+#include "Output.h"
```
Comment (norihiro-w) [inline ProcessLib/Output.cpp]: include at the top

File: ProcessLib/Output.cpp [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+namespace
+{
+//! Determines if there should be output at the given \c timestep.
+template<typename CountsSteps>
+bool doOutput(std::size_t timestep, CountsSteps const& repeats_each_steps)
```
Comment (norihiro-w) [inline ProcessLib/Output.cpp]: i would name it needOutput(). doOutput sounds something is written into a file

File: ProcessLib/Output.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        for (auto pair : timesteps->getConfSubtreeList("pair"))
+        {
+            auto repeat      = pair.getConfParam<std::size_t>("repeat");
+            auto each_steps = pair.getConfParam<std::size_t>("each_steps");
+
+            assert(repeat != 0 && each_steps != 0);
+            out->_repeats_each_steps.emplace_back(repeat, each_steps);
+        }
+
+        if (out->_repeats_each_steps.size() == 0) {
```
Comment (norihiro-w) [inline ProcessLib/Output.cpp]: recommend to use `empty()`

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h (hunk 6 above)
Comment (chleh) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: But this way the time loop can be constructed independently from the project.
Also: It has been like that before this PR.

File: FileIO/VtkIO/PVDFile.cpp (hunk 7 above)
Comment (chleh) [inline FileIO/VtkIO/PVDFile.cpp]: I don't see any problem.

> Destructor should be used only for finalizing resources.

It only finalizes the XML data.

> Moreover, it's not an intuitive interface. A user of this class has to delete an object to get a complete output.

I think, there's no problem. Actually: That's rather intuitive. Since objects are destructed automatically, this implementation asserts that you cannot forget to call some `close()` routine.

**Comment by chleh** [comment]
@norihiro-w Regarding your fundamental objections against this PR: Will you be in the office today? Otherwise I cannot discuss that with you for two weeks, now.

One plea: If you have very substantial comments to a PR (as you have here), please just come over to my office and discuss them. I think that is much easier and more successful than having a discussion on github.

File: ProcessLib/Output.cpp (hunk 10 above)
Comment (chleh) [inline ProcessLib/Output.cpp]: Renamed to `shallDoOutput()`.

File: FileIO/VtkIO/PVDFile.cpp (hunk 8 above)
Comment (chleh) [inline FileIO/VtkIO/PVDFile.cpp]: I pulled it further to the top. But it must be after <fstream> since the ofstream is only forward-declared in PVDFile.h.

File: ProcessLib/Output.cpp (hunk 9 above)
Comment (chleh) [inline ProcessLib/Output.cpp]: Also: must come after <fstream>

**Comment by norihiro-w** [comment]
I'm not sure if I can come to the office today. My daughter is sick and has to be taken care. I have just brought her to a doctor.

**Comment by norihiro-w** [comment]
@chleh I don't think I can work on this today. If you really think merging this PR is urgent, please discuss with others, e.g. @endJunction. If they think no further modification is required, go ahead.

**Comment by chleh** [comment]
I hope your daughter will get well soon.

File: Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h (hunk 6 above)
Comment (norihiro-w) [inline Applications/ApplicationsLib/UncoupledProcessesTimeLoop.h]: i see

File: FileIO/VtkIO/PVDFile.cpp (hunk 7 above)
Comment (norihiro-w) [inline FileIO/VtkIO/PVDFile.cpp]: From my point of view, the XML data is not resource, rather outcome of this functionality. 

I think the interface is not intuitive because usually we (maybe only me?) do not expect a destructor doing something beyond resource finalization (completing the XML data is not included here). Besides, one needs to write the following codes to make it sure the XML is complete at some point. 

```
{
PVDFile pvd("test.pvd");
pvd.addVTUFile();
}

// do something using the PVD file
```

The following codes looks more readable 
```
PVDFile pvd("test.pvd");
pvd.addVTUFile();

// do something using the PVD file
```

or 

```
PVDFile pvd("test.pvd");
pvd.addVTUFile();
pvd.close();

// do something using the PVD file
```

File: ProcessLib/Output.cpp [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+{
+    auto spd_it = _single_process_data.find(&process);
+    if (spd_it == _single_process_data.end()) {
+        ERR("The given process is not contained in the output configuration."
+            " Aborting.");
+        std::abort();
+    }
+    auto& spd = spd_it->second;
+
+    std::string const output_file_name =
```
Comment (norihiro-w) [inline ProcessLib/Output.cpp]: I changed my mind and got idea that it might be better to remove a time step number from the VTU file name, i.e. only including only time information. There is a problem in ogs5 benchmarking that benchmarks using adaptive time stepping can produce result files with different names because the file name includes a time step number and the number can change when someone changed numerics.

File: FileIO/VtkIO/PVDFile.cpp [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <fstream>
+#include "PVDFile.h"
```
Comment (wenqing) [inline FileIO/VtkIO/PVDFile.cpp]: ``#include "PVDFile.h"`` should appear as the first including.

**Comment by norihiro-w** [comment]
@chleh I sent a PR to your branch, which improves PVDFile class and change a VTU file name.

File: ProcessLib/Output.cpp (hunk 12 above)
Comment (chleh) [inline ProcessLib/Output.cpp]: I think the timestep number should be kept in the output file name since it guarantees unique file names; in contrast `std::to_string(t)` does not (probably it will give unique file names for all practical situations, but what if 6 decimal places don't suffice?).
Regarding your adaptive timestepping argument: Probably there should be different timestep numbers only if we change the adaptivity algorithm; and if we do so, not only the name but also the contents of the files will change. Thus, a change in timestep number would mean a rather fundamental change in our code and there will be manual work anyway, so we can also change the timestep numbers in the filename. Or did you get different timestep numbers with the same configuration on different machines? If so: Might that point to a serious problem, since adaptivity is not stable when changing machines?
Furthermore: The output is currently controlled by timestep number, not by time, so changing the former would also change tha latter. Sure, for adaptivity that will have to change (in a future PR).
To summarize: I think we should keep the timestep number for the time being.

**Comment by chleh** [comment]
@norihiro-w Thanks for your PR-PR. It pretty much reflects what you've written in your comments above :smile:
I will merge it when I'm back from vacation.

File: ProcessLib/Output.cpp (hunk 12 above)
Comment (norihiro-w) [inline ProcessLib/Output.cpp]: no we didn't change the algorithm. Whatever changes (bugfix, change of arithmetic operation orders) in assembly can cause different evolution of time stepping.

**Comment by norihiro-w** [comment]
@chleh No further comments from me at the moment. Remaining issues (mesh-dependent output and the naming of VTU files) can be fixed later. From next time, please make a PR smaller, focusing only one feature or a topic for discussion in each PR. It makes reviewing and discussions easier and better.

**Comment by norihiro-w** [comment]
i will check jenkins tomorrow

**Comment by chleh** [comment]
All major discussions settled, now. So, please review this.

**Comment by norihiro-w** [comment]
:+1 from me

File: ProcessLib/Output.cpp [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+            _output_file_prefix + "_pcs_" + std::to_string(spd.process_index)
+            + "_ts_" + std::to_string(timestep)
+            + "_t_"  + std::to_string(t)
+            + ".vtu";
+    DBUG("output to %s", output_file_name.c_str());
+    process.output(output_file_name, timestep, x);
```
Comment (endJunction) [inline ProcessLib/Output.cpp]: Not important notice: In Process::output() the timestep is an unsigned int... Could be consistent

**Comment by endJunction** [comment]
I used this branch together with the time loop PR in M-process and it worked.

File: ProcessLib/Output.cpp (hunk 14 above)
Comment (chleh) [inline ProcessLib/Output.cpp]: Changed the signatures to `unsigned` in the `Output` class.

**Comment by endJunction** [comment]
Jenkins, test this please.

**Comment by endJunction** [comment]
@chleh The ufz/ogs/master needs a reference to the new ufz/ogs-data/master commit for Test/Data submodule. Now it's pointing to ufz/ogs-data/changed-process-output-config...

**Comment by endJunction** [comment]
@chleh Now some of the ctest's are failing. Please have a look into https://svn.ufz.de:8443/job/OGS-6/job/Tests_Linux_Large/318/
There are additional commits on top, but touching only the Changelog.
